core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -995,6 +996,34 @@ private[spark] class AppStatusListener(
     }
   }
 
+  private def updateBroadcastBlock(event: SparkListenerBlockUpdated,
+     broadcast: BroadcastBlockId): Unit = {
+    val now = System.nanoTime()
+    val executorId = event.blockUpdatedInfo.blockManagerId.executorId
+    val storageLevel = event.blockUpdatedInfo.storageLevel
+
+    // Whether values are being added to or removed from the existing accounting.
+    val diskDelta = event.blockUpdatedInfo.diskSize * (if (storageLevel.useDisk) 1 else -1)
+    val memoryDelta = event.blockUpdatedInfo.memSize * (if (storageLevel.useMemory) 1 else -1)
+
+    // Function to apply a delta to a value, but ensure that it doesn't go negative.
+    def newValue(old: Long, delta: Long): Long = math.max(0, old + delta)
+
+    val maybeExec = liveExecutors.get(executorId)
+    maybeExec.foreach { exec =>
+      if (exec.hasMemoryInfo) {
```

Comment:
This block exists in a very similar form in two other places. Feels like time to have a helper method.